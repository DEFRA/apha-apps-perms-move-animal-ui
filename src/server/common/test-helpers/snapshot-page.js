import { parseDocument } from '../../common/test-helpers/dom.js'
import { createServer } from '~/src/server/index.js'
import { withCsrfProtection } from './csrf.js'
/** @import { Server } from '@hapi/hapi' */

/* global expect, it, beforeAll, afterAll, describe */

/**
 * @param {string} describeMessage
 * @param {string} shouldMessage
 * @param {string} urlPath
 */
export const describePageSnapshot = (
  describeMessage,
  shouldMessage,
  urlPath
) => {
  describe(describeMessage, () => {
    /** @type {Server} */
    let server

    beforeAll(async () => {
      server = await createServer()
      await server.initialize()
    })

    afterAll(async () => {
      await server.stop({ timeout: 0 })
    })

    it(shouldMessage, async () => {
      const { payload } = await server.inject(
        withCsrfProtection({
          method: 'GET',
          url: urlPath
        })
      )

      const content =
        parseDocument(payload).querySelector('#main-content')?.innerHTML

      expect(content).toMatchSnapshot()
    })
  })
}
