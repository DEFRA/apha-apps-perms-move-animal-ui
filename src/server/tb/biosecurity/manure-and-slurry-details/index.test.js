import { describePageSnapshot } from '../../../common/test-helpers/snapshot-page.js'
import { disinfectantPage } from '../disinfectant/index.js'
import { slurryManureOtherPage } from '../slurry-manure-other/index.js'
import {
  manureAndSlurryDetailsPage,
  ManureAndSlurryDetailsPage,
  Answer
} from './index.js'

const sectionKey = 'biosecurity'
const question = 'How will you manage slurry and manure?'
const questionKey = 'manureAndSlurryDetails'
const view = 'common/model/page/question-page.njk'
const pageUrl = '/biosecurity/manure-and-slurry-details'

describe('Answer', () => {
  it('should have the correct hint', () => {
    expect(Answer.config.hint).toBe(
      'Only select the measures you are taking. You do not need to select them all to receive your licence.'
    )
  })

  it('should have the correct options', () => {
    expect(Answer.config.options['not-purchased'].label).toBe(
      'Manure is not purchased from other farms'
    )
    expect(Answer.config.options.stored.label).toBe(
      "Manure and slurry is securely stored so domestic and wild animals can't access it"
    )
    expect(Answer.config.options['three-weeks'].label).toBe(
      'Manure is stored for at least 3 weeks before spreading on pasture'
    )
    expect(Answer.config.options['six-months'].label).toBe(
      'Slurry is stored for at least 6 months before spreading on pasture'
    )
    expect(Answer.config.options.other.label).toBe('Other measures')
  })

  it('should have validation for empty input', () => {
    expect(Answer.config.validation.empty?.message).toBe(
      'Enter how you will manage manure and slurry'
    )
  })
})

describe('manureAndSlurryDetails', () => {
  const page = manureAndSlurryDetailsPage

  it('should have the correct urlPath', () => {
    expect(page.urlPath).toBe(pageUrl)
  })

  it('should have the correct sectionKey', () => {
    expect(page.sectionKey).toBe(sectionKey)
  })

  it('should have the correct question', () => {
    expect(page.question).toBe(question)
  })

  it('should have the correct questionKey', () => {
    expect(page.questionKey).toBe(questionKey)
  })

  it('should have the correct view', () => {
    expect(page.view).toBe(view)
  })

  it('should have the correct Answer model', () => {
    expect(page.Answer).toBe(Answer)
  })

  it('nextPage should return disinfectant page', () => {
    const nextPage = page.nextPage({
      value: ['stored']
    })
    expect(nextPage).toBe(disinfectantPage)
  })

  it('nextPage should return slurryManureOtherPage', () => {
    const nextPage = page.nextPage({
      value: ['other']
    })
    expect(nextPage).toBe(slurryManureOtherPage)
  })

  it('should export page', () => {
    expect(manureAndSlurryDetailsPage).toBeInstanceOf(
      ManureAndSlurryDetailsPage
    )
  })

  describePageSnapshot({
    describes: 'manureAndSlurryDetailsPage.content',
    it: 'should render expected response and content',
    pageUrl
  })
})
