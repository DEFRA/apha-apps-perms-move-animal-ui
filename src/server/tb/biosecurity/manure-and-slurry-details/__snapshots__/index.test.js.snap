// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`manureAndSlurryDetails manureAndSlurryDetailsPage.content should render expected response and content 1`] = `
"
    <form method="POST" novalidate="">
      <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
          <input type="hidden" name="crumb" value="csrf-value">
          <input type="hidden" name="nextPage" value="">

<div class="govuk-form-group">
  <fieldset class="govuk-fieldset" aria-describedby="manureAndSlurryDetails-hint">
    <legend class="govuk-fieldset__legend govuk-fieldset__legend--l">
      <h1 class="govuk-fieldset__heading">
        How will you manage slurry and manure?
      </h1>
    </legend>
    <div id="manureAndSlurryDetails-hint" class="govuk-hint">
      Only select the measures you are taking. You do not need to select them all to receive your licence.
    </div>
    <div class="govuk-checkboxes" data-module="govuk-checkboxes">
      <div class="govuk-checkboxes__item">
        <input class="govuk-checkboxes__input" id="manureAndSlurryDetails" name="manureAndSlurryDetails" type="checkbox" value="not-purchased" data-testid="not-purchased-checkbox">
        <label class="govuk-label govuk-checkboxes__label" for="manureAndSlurryDetails">
          Manure is not purchased from other farms
        </label>
      </div>
      <div class="govuk-checkboxes__item">
        <input class="govuk-checkboxes__input" id="manureAndSlurryDetails-2" name="manureAndSlurryDetails" type="checkbox" value="stored" data-testid="stored-checkbox">
        <label class="govuk-label govuk-checkboxes__label" for="manureAndSlurryDetails-2">
          Manure and slurry is securely stored so domestic and wild animals can't access it
        </label>
      </div>
      <div class="govuk-checkboxes__item">
        <input class="govuk-checkboxes__input" id="manureAndSlurryDetails-3" name="manureAndSlurryDetails" type="checkbox" value="three-weeks" data-testid="three-weeks-checkbox">
        <label class="govuk-label govuk-checkboxes__label" for="manureAndSlurryDetails-3">
          Manure is stored for at least 3 weeks before spreading on pasture
        </label>
      </div>
      <div class="govuk-checkboxes__item">
        <input class="govuk-checkboxes__input" id="manureAndSlurryDetails-4" name="manureAndSlurryDetails" type="checkbox" value="six-months" data-testid="six-months-checkbox">
        <label class="govuk-label govuk-checkboxes__label" for="manureAndSlurryDetails-4">
          Slurry is stored for at least 6 months before spreading on pasture
        </label>
      </div>
      <div class="govuk-checkboxes__item">
        <input class="govuk-checkboxes__input" id="manureAndSlurryDetails-5" name="manureAndSlurryDetails" type="checkbox" value="other" data-testid="other-checkbox">
        <label class="govuk-label govuk-checkboxes__label" for="manureAndSlurryDetails-5">
          Other measures
        </label>
      </div>
    </div>
  </fieldset>
</div>

            <button type="submit" class="govuk-button" data-module="govuk-button" id="continue-button">
  Continue
</button>
        </div>
      </div>
    </form>
        "
`;
